### OutFLANK performance on Cichlid Data 

# 2017-09-14

## Packages and Data ##
```{r}
#devtools::install_github("privefl/bigsnpr")
library(OutFLANK)  # outflank package
library(bigsnpr)   # package for LD pruning
library(bigstatsr) # package for LD pruning 
library(pcadapt)

cic_rds<-readRDS("cichlid_data/cichlid.rds") # Note setwd and change path accordingly
cic_met<-read.csv("cichlid_data/cichlid_sample_metadata.csv") # Note setwd and change path accordingly

cic_rds$chromosome_int<-as.integer(as.factor(cic_rds$chromosome))
SNPdata <- t(cic_rds$G)
locinames <- paste(as.integer(cic_rds$fishing_pressure), cic_rds$position, sep="_")
locinames_update <- data.frame(a = cic_rds$chromosome_int, b = cic_rds$position)
```


## OutFLANK analysis with all SNPs
```{r}
FstDataFrame <- MakeDiploidFSTMat(SNPdata,locinames,cic_met$fishing_pressure)
FstDataFrame$PlotPos <- 1:nrow(FstDataFrame)
FstDataFrame$Chrom<- cic_rds$chromosome_int

## FST Correlation
plot(FstDataFrame$FST, FstDataFrame$FSTNoCorr, 
     xlim=c(-0.01,0.3), ylim=c(-0.01,0.3),
     pch=20)
abline(0,1)

k <- 3 ## Number of pops 
out_ini <- OutFLANK(FstDataFrame, NumberOfSamples=k) ## Run outflank on FST dataframe

# Plot results to compare chi-squared distribution vs. actual FST distribution
OutFLANKResultsPlotter(out_ini, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         FALSE, RightZoomFraction = 0.05, titletext = NULL)
## Poor fit, particularly on right tail
OutFLANKResultsPlotter(out_ini, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         TRUE, RightZoomFraction = 0.15, titletext = NULL)

# Histogram of P-values weird
hist(out_ini$results$pvaluesRightTail,breaks = 20)


#### LD Pruning ####
G<-add_code256(big_copy(t(cic_rds$G),type="raw"),code=bigsnpr:::CODE_012)
newpc<-snp_autoSVD(G=G,infos.chr =cic_rds$chromosome_int,infos.pos = cic_rds$position)
plot(newpc) # 3 k groups
which_pruned <- attr(newpc, which="subset") # Indexes of remaining SNPS after pruning


#### Evaluating OutFLANK with pruned data ####
out_trim <- OutFLANK(FstDataFrame[which_pruned,], NumberOfSamples=3)
head(out_trim$results)

OutFLANKResultsPlotter(out_trim, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         FALSE, RightZoomFraction = 0.15, titletext = NULL)
OutFLANKResultsPlotter(out_trim, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         TRUE, RightZoomFraction = 0.10, titletext = NULL)
# Decent distribution fit, no trimming needed. 
hist(out_trim$results$pvaluesRightTail,breaks = 20)
  # still a conservative histogram, which is typical of outflank

top_candidates <- out_trim$results$qvalues<0.01 
# outliers identified conservatively based on qvalue less than 0.01
```

## Plotting
```{r}
# Plot FST for all SNPs (pre-pruning)
plot(FstDataFrame$PlotPos, FstDataFrame$FST,
     col=grey((FstDataFrame$Chrom%%2+1)/3),
     pch=20, main = "Pre-pruning (with post-pruned outliers)")
points(out_trim$results$PlotPos[top_candidates],  out_trim$results$FST[top_candidates], 
       pch=21, cex=2, col=2)

# Simple manhattan plot with only indexed SNPs 
# Note this is does not represent positions on the contigs, simply a way of roughly visuallizing
# relative position of outlier SNPs at each of the three contigs
plot(out_trim$results$PlotPos, out_trim$results$FST,
     col=grey(((locinames_update[which_pruned,]$a)%%2+1)/3),
     pch=20, main="Post-pruning")
# Circle outlier SNPs 
points(out_trim$results$PlotPos[top_candidates],  out_trim$results$FST[top_candidates], 
       pch=21, cex=2, col=2)
```

## PCA structure before pruning
```{r}
res0<-pcadapt(cic_rds$G,K=3)
plot(res0,option="scores",pop=cic_met$place)
```

## PCA structure after pruning
```{r}
dim(cic_rds$G)
res<-pcadapt(cic_rds$G[which_pruned,],K=3)
plot(res,option="scores",pop=cic_met$place)
  # note 2 related individuals
rm <- which(res$scores[,1] > 0.4)
rm
res2<-pcadapt(cic_rds$G[which_pruned, -rm ],K=3)
plot(res2,option="scores",pop=cic_met$place[-rm], xlim=c(-0.2,0.3))
```
